File: src/main/java/com/numberbox/observer/LoginSuccessEventListener.java
package com.numberbox.observer;

import com.numberbox.jwt.entity.RefreshTokenInfo;
import com.numberbox.jwt.repository.RefreshTokenInfoRepository;
import com.numberbox.members.repository.MembersRepository;
import com.numberbox.security.dto.LoginSuccessEvent;
import jakarta.transaction.Transactional;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.UUID;

/**
 * 로그인 성공 후처리
 */
@Component
public class LoginSuccessEventListener {
    private static final int FAIL_CNT_ZERO = 0;
    private static final int HUMAN_STATUS_ENABLE = 0;

    private final MembersRepository membersRepo;
    private final RefreshTokenInfoRepository refreshTokenInfoRepo;

    public LoginSuccessEventListener(MembersRepository membersRepo, RefreshTokenInfoRepository refreshTokenInfoRepo) {
        this.membersRepo = membersRepo;
        this.refreshTokenInfoRepo = refreshTokenInfoRepo;
    }

    @Transactional
    @EventListener
    public void handle(LoginSuccessEvent loginSuccessEvent) {
        final UUID userUniqId = loginSuccessEvent.userId();
        final String refreshToken = loginSuccessEvent.refreshToken();
        final String remainedRefreshToken = loginSuccessEvent.remainedRefreshToken();

        // 로그인 시간, 실패 횟수, 휴면 계정 초기화
        LocalDateTime lastLoginTime = LocalDateTime.now();
        membersRepo.updateByUserUniqId(userUniqId, lastLoginTime, FAIL_CNT_ZERO, HUMAN_STATUS_ENABLE);

        // 기존 refreshToken 남아있는 경우 삭제
        if (remainedRefreshToken != null && !remainedRefreshToken.isEmpty()) {
            refreshTokenInfoRepo.deleteByToken(remainedRefreshToken);
        }

        // 새로운 리프레시 토큰 저장
        RefreshTokenInfo saveToken =
                RefreshTokenInfo.builder().token(refreshToken).userUniqId(userUniqId).build();
        refreshTokenInfoRepo.save(saveToken);
    }
}

File: project/infrastructure/persistence-adapter/src/main/kotlin/com/kamcci/numberbox/infra/persistence/adapter/repository/auth/LoginSuccessEventListener.kt
package com.kamcci.numberbox.infra.persistence.adapter.repository.auth // package com.kamcci.numberbox.infra.persistence.adapter.repository

import com.kamcci.modules.auth.control.dto.LoginSuccessEvent
import com.kamcci.numberbox.infra.orm.jpa.adapter.entity.member.MemberRefreshTokenEntity
import com.kamcci.numberbox.infra.orm.jpa.adapter.repository.member.MemberRefreshTokenRepository
import com.kamcci.numberbox.infra.orm.jpa.adapter.repository.member.MemberRepositorySupport
import org.springframework.context.annotation.Primary
import org.springframework.context.event.EventListener
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.util.*

/**
 * 로그인 성공 후처리
 */
@Primary
@Component
class LoginSuccessEventListener(
    private val memberRepository: MemberRepositorySupport,
    private val memberRefreshTokenRepo: MemberRefreshTokenRepository,
) {

    companion object {
        private const val FAIL_CNT_ZERO = 0
        private const val HUMAN_STATUS_ENABLE = 0
    }

    @Async
    @Transactional
    @EventListener
    fun handle(loginSuccessEvent: LoginSuccessEvent) {
        val userId: UUID = loginSuccessEvent.userId()
        val refreshToken: String = loginSuccessEvent.refreshToken()
        val remainedRefreshToken: String? = loginSuccessEvent.remainedRefreshToken()

        // 실패 횟수, 휴면 계정 초기화
        memberRepository.updateSuccessUser(userId, FAIL_CNT_ZERO, HUMAN_STATUS_ENABLE)

        // 기존 refreshToken 남아있는 경우 삭제
        remainedRefreshToken
            .takeIf { !it.isNullOrEmpty() }
            ?.let {
                memberRefreshTokenRepo.deleteByToken(it)
            }

        // 새로운 리프레시 토큰 저장
        val refreshTokenEntity = MemberRefreshTokenEntity().apply {
            token = refreshToken
            this.memberId = userId
        }
        memberRefreshTokenRepo.save(refreshTokenEntity)
    }
}
